### PR TITLE
gh-112301: Compiler warning management tooling

### PR DESCRIPTION
**Pull request by nohlson**
#112301: This is the first version of compiler warning check tooling integrated into the Ubuntu build ant test GitHub action job.

The tool parses compiler output and checks against an ignore file, which is a list of files for which warnings should be ignored. The script includes options to fail on regression, or fail on improvement (if a file is in the ignore list but the compiler emitted no warnings).

A more detailed writeup about the idea behind the tool can be found [here](http://nohlson.com/blog/Updates-and-First-Tool/)

The current configuration in `reusable-ubuntu.yml` does not include flags to `--fail-on-regression` or `--fail-on-improvement` and simply writes to the log warnings that were emitted but not ignored, or file names that were unexpectedly missing warnings but will not fail the step in the job. Per discussion in the issue we can eventually enable those options.

Future improvements would include removal of duplicate warnings, warning count per file tracking, and adding GitHub Actions annotations.

* Issue: gh-112301

### CODE REVIEW

File: Tools/build/check_warnings.py
Comment (hugovk) [inline Tools/build/check_warnings.py]: What a regex! Please could you attach an example JSON file?

File: Tools/build/check_warnings.py
Comment (hugovk) [inline Tools/build/check_warnings.py]: Please could you limit lines to 79 chars?

https://peps.python.org/pep-0008/#maximum-line-length

This file is in `Tools/`, I think it's okay to just run Black on it:

```
black --line-length 79 Tools/build/check_warnings.py
```

File: Misc/NEWS.d/next/Tests/2024-07-13-21-55-58.gh-issue-112301.YJS1dl.rst
Comment (hugovk) [inline Misc/NEWS.d/next/Tests/2024-07-13-21-55-58.gh-issue-112301.YJS1dl.rst:1]: If you like, you can append "Patch by <Your Name>." here.

**Comment by hugovk** [comment]
(I updated the branch from `main` to fix the unrelated docs error: `Error: new NEWS nits: /home/runner/work/cpython/cpython/build/NEWS:72: WARNING: py:func reference target not found: warnings.filterswarnings`.)

File: Tools/build/check_warnings.py
Comment (hugovk) [inline Tools/build/check_warnings.py]: Let's sort 'em:
```suggestion
import argparse
import json
import re
import sys
from pathlib import Path
```

File: .github/workflows/reusable-ubuntu.yml
Comment (hugovk) [inline .github/workflows/reusable-ubuntu.yml:83]: No warnings so far, is that correct?

https://github.com/python/cpython/actions/runs/9926876262/job/27421126508?pr=121730#step:18:2

File: Tools/build/check_warnings.py
Comment (hugovk) [inline Tools/build/check_warnings.py]:
```suggestion
                    file = file.lstrip('./') # Remove leading current directory if present
```

File: Tools/build/check_warnings.py
Comment (hugovk) [inline Tools/build/check_warnings.py]:
```suggestion
                    file = file.lstrip('./') # Remove leading current directory if present
```

**Review by hugovk — COMMENTED** [review]
Looking good! Here's a few comments and questions.

File: Tools/build/check_warnings.py
Comment (sethmlarson) [inline Tools/build/check_warnings.py:67]: I see a lot of overlap between `get_unexpected_warnings` and `get_unexpected_improvements`, should they be combined somehow or the warning gathering abstracted out?

File: Tools/build/check_warnings.py
Comment (sethmlarson) [inline Tools/build/check_warnings.py:40]: Is there a reason we want to skip malformed JSON?

File: Tools/build/check_warnings.py
Comment (sethmlarson) [inline Tools/build/check_warnings.py]: Since `status` can only be `1`, should we skip the `exit_code |=` and instead `return 1`? These should be put on their own line inside the `if` block as well.

**Review by sethmlarson — COMMENTED** [review]
Thanks @hugovk for reviewing too, I have some more comments:

File: Tools/build/check_warnings.py
Comment (nohlson) [reply Tools/build/check_warnings.py:40]: I found that sometimes the script will grab non-json objects in-between brackets so this just skips those if they can't be interpreted as json

File: Tools/build/check_warnings.py
Comment (nohlson) [reply Tools/build/check_warnings.py]: I wanted the get unexpected warnings and the get unexpected improvements to both run regardless if one fails. I can put the setting the status on a new line

File: .github/workflows/reusable-ubuntu.yml
Comment (nohlson) [reply .github/workflows/reusable-ubuntu.yml:83]: Correct. There are no warnings. I tested the script with warnings and in the blog post I linked in the description I show those tests

File: Tools/build/check_warnings.py
Comment (nohlson) [reply Tools/build/check_warnings.py]: Initially I had a function that used a stack to parse out json objects looking for opening and closing brackets, but to speed things up used a language model with example input and expected output to generate this regex.

Attached is a sample of compiler output with warnings.
[compiler_output.txt](https://github.com/user-attachments/files/16258790/compiler_output.txt)

File: Tools/build/check_warnings.py
Comment (nohlson) [reply Tools/build/check_warnings.py:67]: That is a good point. I refactored the code to extract the filename from the warnings in a separate function. This is only done once and passed to the two routines to check for regressions and improvements

File: Misc/NEWS.d/next/Tests/2024-07-13-21-55-58.gh-issue-112301.YJS1dl.rst
Comment (nohlson) [reply Misc/NEWS.d/next/Tests/2024-07-13-21-55-58.gh-issue-112301.YJS1dl.rst:1]: Added patch by

File: Tools/build/check_warnings.py
Comment (nohlson) [reply Tools/build/check_warnings.py]: Changed comment

File: Tools/build/check_warnings.py
Comment (nohlson) [reply Tools/build/check_warnings.py]: Changed comment

File: Tools/build/check_warnings.py
Comment (nohlson) [reply Tools/build/check_warnings.py]: Sorted imports

File: Tools/build/check_warnings.py
Comment (nohlson) [reply Tools/build/check_warnings.py]: Broke this line into two

**Comment by nohlson** [comment]
I have made the requested changes; please review again

File: Tools/build/check_warnings.py
Comment (sethmlarson) [reply Tools/build/check_warnings.py]: Edit: Removed my suggestion, the regex that I suggested doesn't work. We can keep the current pattern.

File: Tools/build/check_warnings.py
Comment (sethmlarson) [inline Tools/build/check_warnings.py:40]: If we have a stricter regex like above, do we need to keep this step?

**Review by sethmlarson — COMMENTED** [review]
Few more comments, otherwise looks good for the first PR :)

File: Tools/build/check_warnings.py
Comment (nohlson) [reply Tools/build/check_warnings.py:40]: > re.findall(r"^(\[{\"kind\".+\])$", compiler_output, flags=re.MULTILINE|re.DOTALL)

This regex looks like it requires the json arrays to be on a single line and gcc outputs the json arrays across multiple lines.

Were you able to extract the json arrays from the [sample gcc compiler output](https://github.com/python/cpython/pull/121730#discussion_r1680356865)

I am sure there is a more elegant way to grab them but for now the current regex successfully grabs all 156 warnings from the compiler output (many are duplicates)

**Review by sethmlarson — APPROVED** [review]

**Comment by hugovk** [comment]
Hmm, there's a build failure, please can you check it?

```
configure: loading cache config.cache
configure: error: `CFLAGS' was not set in the previous run
configure: error: in `/home/runner/work/cpython/cpython-builddir':
configure: error: changes in the environment can compromise the build
configure: error: run `make distclean' and/or `rm config.cache'
	    and start over
```

https://github.com/python/cpython/actions/runs/10066126406/job/27873774117?pr=121730

**Comment by nohlson** [comment]
> Hmm, there's a build failure, please can you check it?
> 
> ```
> configure: loading cache config.cache
> configure: error: `CFLAGS' was not set in the previous run
> configure: error: in `/home/runner/work/cpython/cpython-builddir':
> configure: error: changes in the environment can compromise the build
> configure: error: run `make distclean' and/or `rm config.cache'
> 	    and start over
> ```
> 
> https://github.com/python/cpython/actions/runs/10066126406/job/27873774117?pr=121730

It looks like for caching reasons I need to include the compiler option to output to json for the `build_ubuntu_ssltests` as well

**Comment by nohlson** [comment]
I have made the requested changes; please review again

File: Misc/NEWS.d/next/Tests/2024-07-13-21-55-58.gh-issue-112301.YJS1dl.rst
Comment (hugovk) [inline Misc/NEWS.d/next/Tests/2024-07-13-21-55-58.gh-issue-112301.YJS1dl.rst]:
```suggestion
Patch by Nate Ohlson.
```

File: Tools/build/check_warnings.py
Comment (hugovk) [inline Tools/build/check_warnings.py:13]: This looks almost good to merge, just some style nits :)

Technically we're outside the stdlib so strictly speaking [PEP 8](https://peps.python.org/pep-0008/) doesn't apply, but let's follow it anyway. For example, please could you add extra newlines before these defs and wrap things to 79 chars?

The easiest way is to run Black (which we don't run on the stdlib but again we're outside the stdlib :)

```
black --line-length 79 Tools/build/check_warnings.py
```
Then some comments and long strings will need manual wrapping.

```suggestion

def extract_warnings_from_compiler_output(compiler_output: str) -> list[dict]:
```

Also, please don't force push in this repo, it makes it easier to review new commits, and we squash merge everything at the end anyway. I think force-pushing is sometimes to blame for pinging lots of unrelated CODEOWNERS too.

> In order to keep the commit history intact, please avoid squashing or amending history and then force-pushing to the PR. Reviewers often want to look at individual commits.

https://devguide.python.org/getting-started/pull-request-lifecycle/#quick-guide

Thanks!

File: Misc/NEWS.d/next/Tests/2024-07-13-21-55-58.gh-issue-112301.YJS1dl.rst
Comment (nohlson) [reply Misc/NEWS.d/next/Tests/2024-07-13-21-55-58.gh-issue-112301.YJS1dl.rst]: Added period

File: Tools/build/check_warnings.py
Comment (nohlson) [reply Tools/build/check_warnings.py:13]: Updated to adhere to PEP8.

After a rebase there were several commits that came from branches from different forks. Not really sure how that happened but I had to revert to my last commit and force push to get rid of them.

Comment (hugovk) [reply Tools/build/check_warnings.py:13]: Thank you! I just pushed a commit to also wrap comments and strings to 79 chars as well.

**Review by hugovk — APPROVED** [review]
Thanks!

File: Tools/build/check_warnings.py
Comment (picnixz) [inline Tools/build/check_warnings.py:117]: Can we have an alias as well? in case we want to use it ourselves? (like `-c`).

File: Tools/build/check_warnings.py
Comment (picnixz) [inline Tools/build/check_warnings.py:123]: Same here, like `-i`.

File: Tools/build/check_warnings.py
Comment (picnixz) [inline Tools/build/check_warnings.py:129]: And here as well, like `-x`

File: Tools/build/check_warnings.py
Comment (picnixz) [inline Tools/build/check_warnings.py:135]: And here as well, like `-X`.

File: Tools/build/check_warnings.py
Comment (picnixz) [inline Tools/build/check_warnings.py:1]: Do we need this one?

Comment (hugovk) [reply Tools/build/check_warnings.py:1]: I meant to comment this too: no, I don't think we do because the file does not have the executable bit and we're running it via `python Tools/build/check_warnings.py`

**Comment by hugovk** [comment]
@picnixz Your review comments came after I'd enabled auto-merge, sorry about that!

---

@nohlson Hi! @corona10 noticed that with this change, the build output is no longer being output to the logs, because we're redirecting it to a file. It can sometimes be useful to see this output.

@zware suggested a command like this:

```sh
set -o pipefail; make -j4 2>&1 | tee compiler_output.txt
```

Please could you open a PR to try this out? Thank you!

File: Tools/build/check_warnings.py
Comment (picnixz) [reply Tools/build/check_warnings.py:117]: (Adressed in https://github.com/python/cpython/pull/122465)

**Pull request merged**
